tools/run_cargo_fmt.sh

```diff
@@ -29,35 +29,62 @@ fi
 # |M  changed file, staged (git add has run)
 # |MM changed file, some staged and some unstaged changes (git add then changes)
 # |?? untracked file
-if git status --porcelain | grep '^.M.*\.rs' -q; then
-	echo "$(tput bold)Warning: Formatting will overwrite files in place.$(tput sgr0)"
-	echo "While this is probably what you want, it's often useful to"
-	echo "stage all of your changes (git add ...) before format runs,"
-	echo "just so you can double-check everything."
-	echo ""
-	echo "$(tput bold)git status:$(tput sgr0)"
-	git status
-	echo ""
-	read -p "Continue formatting with unstaged changes? [y/N] " response
-	if [[ ! ( "$(echo "$response" | tr :upper: :lower:)" == "y" ) ]]; then
-		exit 0
+if [ "$1" != "diff" ]; then
+	if git status --porcelain | grep '^.M.*\.rs' -q; then
+		echo "$(tput bold)Warning: Formatting will overwrite files in place.$(tput sgr0)"
+		echo "While this is probably what you want, it's often useful to"
+		echo "stage all of your changes (git add ...) before format runs,"
+		echo "just so you can double-check everything."
+		echo ""
+		echo "$(tput bold)git status:$(tput sgr0)"
+		git status
+		echo ""
+		read -p "Continue formatting with unstaged changes? [y/N] " response
+		if [[ ! ( "$(echo "$response" | tr :upper: :lower:)" == "y" ) ]]; then
+			exit 0
+		fi
 	fi
 fi
 
 set +e
 let FAIL=0
 set -e
 
+# Get the list of files in the workspace to avoid formatting
+# them twice
+csplit Cargo.toml '/exclude = \[/' > /dev/null
+rm xx01
+
 # Find folders with Cargo.toml files in them and run `cargo fmt`.
-if [ "$1" == "diff" ]; then
-	# Just print out diffs and count errors, used by Travis
-	CARGO_FMT_ARGS="-- --check"
-fi
 for f in $(find . | grep Cargo.toml); do
```

Comment:
Is there any reason to keep this loop, as opposed to running `cargo fmt` directly on the workspace?

Reply:
I agree we should switch to `cargo fmt --all`.

Reply:
I think this is already doing what we want.

When run in the root of a workspace, `cargo fmt` runs on all modules in the workspace. That is what the `== ./Cargo.toml` case covers. The `--all` flag is only meaningful if you want to run `cargo fmt` from in a module directory but want it to format the whole workspace instead of just the module you're running in (i.e. you typed `cargo fmt` in the `boards/hail` directory).

In no case does `cargo fmt --all` choose to format modules that are listed as excluded from the workspace. The format script first runs `cargo fmt` in the root, which formats the whole workspace, and then the loop finds all `Cargo.toml`s in the tree and skips any that were already formatted as part of the workspace.

I tested this fairly extensively, making sure that nothing is double-formatted nor is anything skipped in formatting. Is there any incorrect behavior you're observing?

Reply:
For me, it is:

```
$ time make fmt
Formatting complete.

********************************************************************************
* Formatting complete                                                          *
********************************************************************************


real	0m10.809s
user	0m3.021s
sys	0m3.869s

 $ time cargo fmt --all

real	0m4.538s
user	0m3.467s
sys	0m0.933s
```

Reply:
It takes longer because it does more. `cargo fmt --all` doesn't format every module in the tock tree:

```
[-bash] Thu 28 May 15:56 [[ci-policy-impl $] ~/code/helena-project/tock]
$ cargo fmt --all -- --check
[-bash] Thu 28 May 15:56 [[ci-policy-impl $] ~/code/helena-project/tock]
$ make ci-job-format

********************************************************************************
* CI-Job: Format Check                                                         *
********************************************************************************

Formatting tools/alert_codes                      <- Contains formatting errors!
Diff in /Volumes/code/helena-project/tock/tools/alert_codes/src/main.rs at line 17:

 /// Returns the alert code specified on the command line, or prints a usage
 /// error if it was omitted or incorrectly specified.
-   fn get_alert_code() -> Result<u32, ()> {
+fn get_alert_code() -> Result<u32, ()> {
     let mut args = std::env::args_os();
     if args.len() != 2 {
         usage_error("Incorrect number of arguments");

Formatting complete.

Formatting errors.
See above for details
make[1]: *** [tools/.format_fresh] Error 1
make: *** [ci-job-format] Error 2
```

There is literally no difference between running `cargo fmt` and `cargo fmt --all` **in the root of the workspace**. The only thing the `--all` flag does is change the behavior when run in a module (aka in a subdirectory).

The script runs `cargo fmt` in the root which is exactly equivalent to running `cargo fmt --all` and then checks the crates that excluded from the cargo workspace as well.

Reply:
I now realize that this version is 2x faster than what is in master now, so that's good. I worry about the maintainability of this script, but `cargo fmt --all` works well enough if we need to fall back to that.

Reply:
I think a good solution to this problem would be to have a second workspace for tools. Then we can run `cargo fmt` twice: in the root directory and in the `tools/` directory. But this is not in scope for this pull request.